File: android/app/src/main/java/com/lockoutprotocol/guardian/focus/FocusSession.kt
package com.lockoutprotocol.guardian.focus

import android.content.Context
import com.lockoutprotocol.guardian.data.Prefs
import java.io.File
import java.util.UUID
import org.json.JSONArray
import org.json.JSONObject

/**
 * The two accountability levels, differing only in who holds the exit.
 *
 * Neither can trap you: closing a blocked app is always passcode-free. What [LOCKED] costs is the
 * override, not the escape. See SAFEGUARDS.md.
 */
enum class Accountability(val id: String) {
    SELF("self"),
    LOCKED("locked");

    val title: String
        get() = when (this) {
            SELF -> "Self-managed"
            LOCKED -> "Locked"
        }

    val blurb: String
        get() = when (this) {
            SELF -> "It blocks you and logs it — but you can wave it away yourself. " +
                "No passcode, nobody is told. Start here."
            LOCKED -> "Overriding a block needs the passcode, and every block and override is " +
                "pushed to your accountability partner. Closing the app still always works."
        }

    val requiresPasscodeToEnd: Boolean get() = this == LOCKED
    val alertsPartner: Boolean get() = this == LOCKED

    companion object {
        /** Unknown values fall back to the weaker mode, never the stronger one. */
        fun from(id: String?): Accountability = entries.firstOrNull { it.id == id } ?: SELF
    }
}

/**
 * One declared stretch of work. When no session is running, no screenshot is taken at all — which
 * matters more here than on the desktop, since the app holds an AccessibilityService.
 */
data class FocusSession(
    val id: String = "fs_" + UUID.randomUUID().toString().take(10).lowercase(),
    val task: String,
    val startedAt: Long = System.currentTimeMillis(),
    /** 0 = open-ended. Not offered for [Accountability.LOCKED]: open-ended plus a forgotten
     *  passcode is the one shape that could genuinely trap someone. */
    val plannedMinutes: Int = 0,
    val intervalSeconds: Int = DEFAULT_INTERVAL,
    val accountability: Accountability = Accountability.SELF,
    val extraApps: Set<String> = emptySet(),
    val allowedApps: Set<String> = emptySet(),
    val providerId: String = "",
    var endedAt: Long? = null,
    var endedReason: String = "",
    var checks: Int = 0,
    var offTaskCount: Int = 0,
    var overrideCount: Int = 0,
    var pausedUntil: Long = 0L,
) {

    /**
     * Saved defaults, plus anything added just for today, minus anything excused just for today.
     * Stored as a delta rather than a copy, so editing the defaults mid-session takes effect.
     */
    fun watchlist(defaults: Set<String>): Set<String> = (defaults + extraApps) - allowedApps

    val isActive: Boolean get() = endedAt == null
    val elapsedMs: Long get() = (System.currentTimeMillis() - startedAt).coerceAtLeast(0)

    /** Milliseconds left, or null for an open-ended session. */
    val remainingMs: Long?
        get() = if (plannedMinutes <= 0) null
        else (plannedMinutes * 60_000L - elapsedMs).coerceAtLeast(0)

    val isOver: Boolean get() = remainingMs?.let { it <= 0 } ?: false
    val isPaused: Boolean get() = pausedUntil > System.currentTimeMillis()

    val remainingText: String
        get() {
            val left = remainingMs ?: return "open-ended"
            return "${left / 60_000}m ${((left % 60_000) / 1000).toString().padStart(2, '0')}s left"
        }

    fun toJson(): JSONObject = JSONObject().apply {
        put("id", id)
        put("task", task)
        put("started_at", startedAt)
        put("planned_minutes", plannedMinutes)
        put("interval_seconds", intervalSeconds)
        put("accountability", accountability.id)
        put("extra_apps", JSONArray(extraApps.toList()))
        put("allowed_apps", JSONArray(allowedApps.toList()))
        put("provider_id", providerId)
        put("ended_at", endedAt ?: JSONObject.NULL)
        put("ended_reason", endedReason)
        put("checks", checks)
        put("off_task_count", offTaskCount)
        put("override_count", overrideCount)
        put("paused_until", pausedUntil)
    }

    companion object {
        /** Below a minute you pay for inference constantly; above five a detour has already eaten
         *  the block it should have prevented. */
        const val DEFAULT_INTERVAL = 120
        const val MIN_INTERVAL = 15
        const val MAX_INTERVAL = 3600

        fun clampInterval(seconds: Int): Int = seconds.coerceIn(MIN_INTERVAL, MAX_INTERVAL)

        private fun stringSet(o: JSONObject, key: String): Set<String> {
            val arr = o.optJSONArray(key) ?: return emptySet()
            return (0 until arr.length()).mapNotNull { arr.optString(it).takeIf(String::isNotEmpty) }
                .toSet()
        }

        fun fromJson(o: JSONObject): FocusSession = FocusSession(
            id = o.optString("id", "fs_unknown"),
            task = o.optString("task", ""),
            startedAt = o.optLong("started_at", System.currentTimeMillis()),
            plannedMinutes = o.optInt("planned_minutes", 0).coerceAtLeast(0),
            intervalSeconds = clampInterval(o.optInt("interval_seconds", DEFAULT_INTERVAL)),
            accountability = Accountability.from(o.optString("accountability")),
            extraApps = stringSet(o, "extra_apps"),
            allowedApps = stringSet(o, "allowed_apps"),
            providerId = o.optString("provider_id", ""),
        ).apply {
            endedAt = if (o.isNull("ended_at")) null else o.optLong("ended_at")
            endedReason = o.optString("ended_reason", "")
            checks = o.optInt("checks", 0)
            offTaskCount = o.optInt("off_task_count", 0)
            overrideCount = o.optInt("override_count", 0)
            pausedUntil = o.optLong("paused_until", 0L)
        }
    }
}

/**
 * The one active session, on disk, plus an append-only history.
 *
 * Persistence is load-bearing: Android kills foreground services, and if a restart silently
 * cancelled a locked session then force-stopping the app would be a one-tap bypass.
 */
object SessionStore {

    private const val SESSION_FILE = "focus_session.json"
    private const val HISTORY_FILE = "focus_history.jsonl"

    private var cached: FocusSession? = null
    private var loaded = false
    private val listeners = mutableListOf<() -> Unit>()

    private fun sessionFile(ctx: Context) = File(ctx.filesDir, SESSION_FILE)
    private fun historyFile(ctx: Context) = File(ctx.filesDir, HISTORY_FILE)

    fun subscribe(fn: () -> Unit) { listeners += fn }

    private fun notifyChanged() = listeners.toList().forEach { runCatching { it() } }

    private fun load(ctx: Context) {
        if (loaded) return
        loaded = true
        cached = runCatching {
            val f = sessionFile(ctx)
            if (!f.exists()) return@runCatching null
            FocusSession.fromJson(JSONObject(f.readText())).takeIf { it.isActive }
        }.getOrNull()
    }

    /** The live session, or null. Expiry is evaluated on read, so a phone in Doze past the end
     *  time still ends its session cleanly. */
    @Synchronized
    fun current(ctx: Context): FocusSession? {
        load(ctx)
        val s = cached ?: return null
        if (s.isOver) {
            end(ctx, "time's up")
            return null
        }
        return cached
    }

    fun isActive(ctx: Context): Boolean = current(ctx) != null

    @Synchronized
    fun start(ctx: Context, session: FocusSession): FocusSession {
        load(ctx)
        if (cached != null) end(ctx, "replaced by a new session")
        cached = session
        save(ctx)
        notifyChanged()
        return session
    }

    @Synchronized
    fun end(ctx: Context, reason: String = "ended") {
        load(ctx)
        val s = cached ?: return
        s.endedAt = System.currentTimeMillis()
        s.endedReason = reason
        runCatching { historyFile(ctx).appendText(s.toJson().toString() + "\n") }
        cached = null
        save(ctx)
        notifyChanged()
    }

    /** Used by the block screen's override, so the user isn't re-blocked mid-sentence. */
    @Synchronized
    fun pause(ctx: Context, millis: Long) {
        load(ctx)
        cached?.let {
            it.pausedUntil = System.currentTimeMillis() + millis.coerceAtLeast(0)
            save(ctx)
            notifyChanged()
        }
    }

    @Synchronized
    fun recordCheck(ctx: Context, offTask: Boolean) {
        load(ctx)
        cached?.let {
            it.checks += 1
            if (offTask) it.offTaskCount += 1
            save(ctx)
        }
    }

    @Synchronized
    fun recordOverride(ctx: Context) {
        load(ctx)
        cached?.let {
            it.overrideCount += 1
            save(ctx)
            notifyChanged()
        }
    }

    private fun save(ctx: Context) {
        runCatching {
            val f = sessionFile(ctx)
            val s = cached
            if (s == null) {
                f.delete()
            } else {
                // Write-then-rename: a kill mid-write must not leave a truncated file, because an
                // unreadable session reads as "no session" and silently unlocks.
                val tmp = File(ctx.filesDir, "$SESSION_FILE.tmp")
                tmp.writeText(s.toJson().toString())
                tmp.renameTo(f)
            }
        }
    }

    /** Finished sessions, newest last. A torn final line after a hard kill is skipped. */
    fun history(ctx: Context, limit: Int = 50): List<FocusSession> {
        val f = historyFile(ctx)
        if (!f.exists()) return emptyList()
        return runCatching {
            f.readLines().takeLast(limit).mapNotNull { line ->
                runCatching { FocusSession.fromJson(JSONObject(line)) }.getOrNull()
            }
        }.getOrDefault(emptyList())
    }

    fun defaultsFor(ctx: Context): Set<String> = Prefs.get(ctx).monitoredPackages
}

File: macos/Guardian/Focus/FocusSession.swift
import Foundation

/// The two accountability levels, differing only in who holds the exit.
///
/// Neither can trap you: closing a blocked app is always passcode-free. What `.locked` costs is
/// the override, not the escape. See `SAFEGUARDS.md`.
enum Accountability: String, Codable, CaseIterable, Sendable {
    case selfManaged = "self"
    case locked = "locked"

    var title: String {
        switch self {
        case .selfManaged: return "Self-managed"
        case .locked: return "Locked"
        }
    }

    var blurb: String {
        switch self {
        case .selfManaged:
            return "It blocks you and logs it — but you can wave it away yourself. "
                 + "No passcode, nobody is told. Start here."
        case .locked:
            return "Overriding a block needs the passcode, and every block and override is pushed "
                 + "to your accountability partner. Closing the app still always works."
        }
    }

    /// Only a locked session holds its own exit; you can always quit the app.
    var requiresPasscodeToEnd: Bool { self == .locked }
    var alertsPartner: Bool { self == .locked }
}

/// One declared stretch of work: a task, a length, a set of apps, checked every
/// `intervalSeconds`. When no session is running, nothing is captured at all.
struct FocusSession: Codable, Identifiable, Sendable {

    /// Below a minute you pay for inference constantly; above five a detour has already eaten
    /// the block it should have prevented.
    static let defaultInterval = 120
    static let minInterval = 15
    static let maxInterval = 3600

    var id: String
    var task: String
    var startedAt: Date
    /// 0 = open-ended. Not offered for `.locked`: open-ended plus a forgotten passcode is the
    /// one shape that could genuinely trap someone.
    var plannedMinutes: Int
    var intervalSeconds: Int
    var accountability: Accountability
    /// This-session-only adjustments over the saved default watchlist.
    var extraApps: Set<String>
    var allowedApps: Set<String>
    var providerId: String
    var endedAt: Date?
    var endedReason: String
    var checks: Int
    var offTaskCount: Int
    var overrideCount: Int
    var pausedUntil: Date?

    init(task: String,
         plannedMinutes: Int = 0,
         intervalSeconds: Int = FocusSession.defaultInterval,
         accountability: Accountability = .selfManaged,
         extraApps: Set<String> = [],
         allowedApps: Set<String> = [],
         providerId: String = "") {
        self.id = "fs_" + UUID().uuidString.prefix(10).lowercased()
        self.task = task.trimmingCharacters(in: .whitespacesAndNewlines)
        self.startedAt = Date()
        self.plannedMinutes = max(plannedMinutes, 0)
        self.intervalSeconds = Self.clampInterval(intervalSeconds)
        self.accountability = accountability
        self.extraApps = extraApps
        self.allowedApps = allowedApps
        self.providerId = providerId
        self.endedAt = nil
        self.endedReason = ""
        self.checks = 0
        self.offTaskCount = 0
        self.overrideCount = 0
        self.pausedUntil = nil
    }

    static func clampInterval(_ seconds: Int) -> Int {
        min(max(seconds, minInterval), maxInterval)
    }

    // MARK: - Watchlist

    /// Saved defaults, plus anything added just for today, minus anything excused just for today.
    /// Stored as a delta rather than a copy, so editing the defaults mid-session takes effect.
    func watchlist(defaults: Set<String>) -> Set<String> {
        defaults.union(extraApps).subtracting(allowedApps)
    }

    // MARK: - Lifecycle

    var isActive: Bool { endedAt == nil }
    var elapsed: TimeInterval { max(Date().timeIntervalSince(startedAt), 0) }

    /// Seconds left, or nil for an open-ended session.
    var remaining: TimeInterval? {
        guard plannedMinutes > 0 else { return nil }
        return max(Double(plannedMinutes * 60) - elapsed, 0)
    }

    var isOver: Bool {
        guard let remaining else { return false }
        return remaining <= 0
    }

    var isPaused: Bool {
        guard let pausedUntil else { return false }
        return pausedUntil > Date()
    }

    var remainingText: String {
        guard let remaining else { return "open-ended" }
        return "\(Int(remaining) / 60)m \(String(format: "%02d", Int(remaining) % 60))s left"
    }
}

/// The one active session, persisted, plus an append-only history.
///
/// Persistence is load-bearing: if force-quitting silently cancelled a locked session, "locked"
/// would be worth nothing.
@MainActor
final class SessionStore: ObservableObject {

    static let shared = SessionStore()

    @Published private(set) var stored: FocusSession?

    private let sessionURL: URL
    private let historyURL: URL

    private init() {
        let dir = Self.supportDirectory()
        sessionURL = dir.appendingPathComponent("focus_session.json")
        historyURL = dir.appendingPathComponent("focus_history.jsonl")
        stored = Self.load(from: sessionURL)
        if let s = stored, !s.isActive { stored = nil }
    }

    private static func supportDirectory() -> URL {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)
            .first ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("LockoutProtocol", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    // MARK: - Accessors

    /// The live session, or nil. Expiry is evaluated on read, so a Mac that slept past the end
    /// time still ends its session cleanly.
    var current: FocusSession? {
        guard let s = stored else { return nil }
        if s.isOver {
            end(reason: "time's up")
            return nil
        }
        return stored
    }

    var isActive: Bool { current != nil }

    // MARK: - Mutations

    @discardableResult
    func start(_ session: FocusSession) -> FocusSession {
        if stored != nil { end(reason: "replaced by a new session") }
        stored = session
        save()
        return session
    }

    func end(reason: String = "ended") {
        guard var s = stored else { return }
        s.endedAt = Date()
        s.endedReason = reason
        archive(s)
        stored = nil
        save()
    }

    /// Used by the block screen's override, so the user isn't re-blocked mid-sentence.
    func pause(seconds: TimeInterval) {
        guard var s = stored else { return }
        s.pausedUntil = Date().addingTimeInterval(max(seconds, 0))
        stored = s
        save()
    }

    func recordCheck(offTask: Bool) {
        guard var s = stored else { return }
        s.checks += 1
        if offTask { s.offTaskCount += 1 }
        stored = s
        save()
    }

    func recordOverride() {
        guard var s = stored else { return }
        s.overrideCount += 1
        stored = s
        save()
    }

    func addSessionApp(_ bundleId: String) {
        guard var s = stored else { return }
        s.extraApps.insert(bundleId)
        s.allowedApps.remove(bundleId)
        stored = s
        save()
    }

    func allowSessionApp(_ bundleId: String) {
        guard var s = stored else { return }
        s.allowedApps.insert(bundleId)
        s.extraApps.remove(bundleId)
        stored = s
        save()
    }

    // MARK: - Persistence

    private static func load(from url: URL) -> FocusSession? {
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? JSONDecoder().decode(FocusSession.self, from: data)
    }

    private func save() {
        guard let s = stored else {
            try? FileManager.default.removeItem(at: sessionURL)
            return
        }
        guard let data = try? JSONEncoder().encode(s) else { return }
        try? data.write(to: sessionURL, options: .atomic)
    }

    private func archive(_ session: FocusSession) {
        guard var data = try? JSONEncoder().encode(session) else { return }
        data.append(contentsOf: [0x0A])                       // newline — this is JSONL
        if let handle = try? FileHandle(forWritingTo: historyURL) {
            defer { try? handle.close() }
            try? handle.seekToEnd()
            try? handle.write(contentsOf: data)
        } else {
            try? data.write(to: historyURL, options: .atomic)
        }
    }

    /// Finished sessions, newest last. A torn final line after a hard kill is skipped.
    func history(limit: Int = 50) -> [FocusSession] {
        guard let text = try? String(contentsOf: historyURL, encoding: .utf8) else { return [] }
        let decoder = JSONDecoder()
        return text.split(separator: "\n").suffix(limit).compactMap {
            guard let data = $0.data(using: .utf8) else { return nil }
            return try? decoder.decode(FocusSession.self, from: data)
        }
    }
}
